bftengine/src/bcstatetransfer/BCStateTran.cpp

```diff
@@ -1385,56 +1419,100 @@ bool BCStateTran::onMessage(const FetchBlocksMsg *m, uint32_t msgLen, uint16_t r
     registrar.perf.snapshot("state_transfer");
     registrar.perf.snapshot("state_transfer_src");
   }
-  // compute information about next block and chunk
-  uint64_t nextBlock = m->lastRequiredBlock;
-  uint32_t sizeOfNextBlock = 0;
-  bool tmp = as_->getBlock(nextBlock, buffer_, &sizeOfNextBlock);
-  ConcordAssert(tmp);
-  ConcordAssertGT(sizeOfNextBlock, 0);
 
-  uint32_t sizeOfLastChunk = config_.maxChunkSize;
-  uint32_t numOfChunksInNextBlock = sizeOfNextBlock / config_.maxChunkSize;
-  if (sizeOfNextBlock % config_.maxChunkSize != 0) {
-    sizeOfLastChunk = sizeOfNextBlock % config_.maxChunkSize;
-    numOfChunksInNextBlock++;
-  }
+  // start recording time to send a whole batch, and its size
+  uint64_t batchSizeBytes = 0;
+  uint64_t batchSizeBlocks = 0;
+  src_send_batch_duration_rec_.clear();
+  src_send_batch_duration_rec_.start();
 
+  // compute information about next block and chunk
+  uint64_t nextBlockId = m->lastRequiredBlock;
   uint16_t nextChunk = m->lastKnownChunkInLastRequiredBlock + 1;
+  uint16_t numOfSentChunks = 0;
 
-  // if msg is invalid (lastKnownChunkInLastRequiredBlock+1 does not exist)
-  if (nextChunk > numOfChunksInNextBlock) {
-    LOG_WARN(getLogger(),
-             "Msg is invalid: illegal chunk number: " << KVLOG(replicaId, nextChunk, numOfChunksInNextBlock));
-    return false;
+  auto isFetchersContextValid = srcGetBlockContextes_[0].future.valid();
+  if (!isFetchersContextValid || srcGetBlockContextes_[0].blockId != nextBlockId) {
+    LOG_INFO(getLogger(),
+             "Call asyncGetBlocksConcurrent: source blocks prefetch disabled (first batch or retransmission): "
```

Comment:
Disabling the feature will is used by param, I will upload a separate PR.
This is the case of 1st fetch or prediction failed.
In both cases, it will fetch all blocks again.